contracts/staking/StakingPool.sol

```diff
@@ -0,0 +1,470 @@
+                );
+            }
+        }
+
+        emit Deposit(_msgSender(), amount);
+
+        require(
+            _info.stakeToken.transferFrom(_msgSender(), address(this), amount),
+            "StakingPool: failed to transfer"
+        );
+    }
+
+    /**
+     * @notice Withdraw both stake and reward tokens when the stake period is complete
+     */
+    function withdraw()
+        external
+        stakingPeriodComplete
+        rewardsAvailable
+        nonReentrant
+    {
+        User memory user = _users[_msgSender()];
+        require(user.depositAmount > 0, "StakingPool: not eligible");
+
+        delete _users[_msgSender()];
+
+        StakingPoolLib.Data storage _info = _stakingPoolInfo;
+
+        //slither-disable-next-line reentrancy-events
+        _transferStake(user.depositAmount, IERC20(_info.stakeToken));
+
+        // calculate the rewardAmounts due to the user
+        for (uint256 i = 0; i < _info.rewardTokens.length; i++) {
+            uint256 amount = 0;
+
+            // floating
+            if (_info.poolType == StakingPoolLib.StakingPoolType.FLOATING) {
+                amount = _calculateFloatingReward(
+                    _info.rewardTokens[i].rewardAmountRatio,
+                    user.depositAmount
+                );
+            } else {
+                // fixed
+                amount = uint256(user.rewardAmounts[i]);
+            }
+            //slither-disable-next-line calls-loop
+            _transferRewards(amount, IERC20(_info.rewardTokens[i].token));
+        }
+    }
+
+    /**
+     * @notice Withdraw only stake tokens. Reward tokens may not be available/unlocked yet.
+     */
+    function withdrawStake() external stakingPeriodComplete nonReentrant {
+        User storage user = _users[_msgSender()];
+        require(user.depositAmount > 0, "StakingPool: not eligible");
+
+        uint128 currentDepositBalance = user.depositAmount;
+        user.depositAmount = 0;
+
+        StakingPoolLib.Data storage _info = _stakingPoolInfo;
+
+        // calculate the amount of rewards the user is due for the floating pool type
+        // fixed amounts are calculated and fixed on depositing
+        for (uint256 i = 0; i < _info.rewardTokens.length; i++) {
+            if (_info.poolType == StakingPoolLib.StakingPoolType.FLOATING) {
+                user.rewardAmounts[i] = _calculateFloatingReward(
+                    _info.rewardTokens[i].rewardAmountRatio,
+                    currentDepositBalance
+                );
+            }
+        }
+
+        _transferStake(currentDepositBalance, IERC20(_info.stakeToken));
+    }
+
+    /**
+     * @notice Withdraw only reward tokens. Stake may have already been withdrawn.
+     */
+    function withdrawRewards() external stakingPeriodComplete rewardsAvailable {
+        User memory user = _users[_msgSender()];
+        require(user.rewardAmounts[0] > 0, "StakingPool: No rewards"); // this is safe?
```

Comment:
question: what is the intent of the comment `// this is safe?`

Don't understand the comment 🤔

Reply:
...wait, why is the index of the rewards array zero?

Reply:
iterate over array length

Reply:
removing